README.md
# it-pushable <!-- omit in toc -->

[![codecov](https://img.shields.io/codecov/c/github/alanshaw/it-pushable.svg?style=flat-square)](https://codecov.io/gh/alanshaw/it-pushable)
[![CI](https://img.shields.io/github/actions/workflow/status/alanshaw/it-pushable/js-test-and-release.yml?branch=master\&style=flat-square)](https://github.com/alanshaw/it-pushable/actions/workflows/js-test-and-release.yml?query=branch%3Amaster)

> An iterable that you can push values into

## Table of contents <!-- omit in toc -->

- [Install](#install)
  - [Browser `<script>` tag](#browser-script-tag)
- [Usage](#usage)
  - [Backpressure](#backpressure)
- [Related](#related)
- [API Docs](#api-docs)
- [License](#license)
- [Contribution](#contribution)

## Install

```console
$ npm i it-pushable
```

### Browser `<script>` tag

Loading this module through a script tag will make it's exports available as `ItPushable` in the global namespace.

```html
<script src="https://unpkg.com/it-pushable/dist/index.min.js"></script>
```

## Usage

```js
import { pushable } from 'it-pushable'

const source = pushable()

setTimeout(() => source.push('hello'), 100)
setTimeout(() => source.push('world'), 200)
setTimeout(() => source.end(), 300)

const start = Date.now()

for await (const value of source) {
  console.log(`got "${value}" after ${Date.now() - start}ms`)
}
console.log(`done after ${Date.now() - start}ms`)

/*
Output:
got "hello" after 105ms
got "world" after 207ms
done after 309ms
*/
```

```js
import { pushableV } from 'it-pushable'
import all from 'it-all'

const source = pushableV()

source.push(1)
source.push(2)
source.push(3)
source.end()

console.info(await all(source))
/*
Output:
[ [1, 2, 3] ]
*/
```

### Backpressure

This module supports backpressure by returning a promise to any `push` or `end`
call.

These promises will resolve when the `push`ed value has been consumed (in the
case of `.push`) or when the consumer has iterated over all values in the
pushable (in the case of `.end`).

If you do not wish to wait and instead buffer as much data as possible, do not
await the result of `.push`

```js
import { pushable } from 'it-pushable'

const source = pushable()

// wait for the value to be consumed
await source.push(5)

// do not wait for the value to be consumed
void source.push(5)

// only wait 10ms for the value to be consumed
await source.push(5, {
  signal: AbortSignal.timeout(10)
})
```

## Related

- [`it-pipe`](https://www.npmjs.com/package/it-pipe) Utility to "pipe" async iterables together

## API Docs

- <https://alanshaw.github.io/it-pushable>

## License

Licensed under either of

- Apache 2.0, ([LICENSE-APACHE](LICENSE-APACHE) / <http://www.apache.org/licenses/LICENSE-2.0>)
- MIT ([LICENSE-MIT](LICENSE-MIT) / <http://opensource.org/licenses/MIT>)

## Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted for inclusion in the work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any additional terms or conditions.


CHANGELOG.md
## [3.2.3](https://github.com/alanshaw/it-pushable/compare/v3.2.2...v3.2.3) (2023-11-16)


### Bug Fixes

* support onEmpty with onEnd ([#76](https://github.com/alanshaw/it-pushable/issues/76)) ([26b5d18](https://github.com/alanshaw/it-pushable/commit/26b5d18d788b16833d66c4063a1c9f07d5a57e53))

## [3.2.2](https://github.com/alanshaw/it-pushable/compare/v3.2.1...v3.2.2) (2023-11-11)


### Trivial Changes

* add or force update .github/workflows/js-test-and-release.yml ([#66](https://github.com/alanshaw/it-pushable/issues/66)) ([9dc3452](https://github.com/alanshaw/it-pushable/commit/9dc3452fb1d6dc7a978dec2b429204996e28eb3a))
* delete templates [skip ci] ([#65](https://github.com/alanshaw/it-pushable/issues/65)) ([623e049](https://github.com/alanshaw/it-pushable/commit/623e0496897df1db83c5677c3540b5b708251372))


### Dependencies

* **dev:** bump aegir from 39.0.13 to 41.0.5 ([#74](https://github.com/alanshaw/it-pushable/issues/74)) ([ca99910](https://github.com/alanshaw/it-pushable/commit/ca9991015a1852c2aedc16af149237ff5c083742))

## [3.2.1](https://github.com/alanshaw/it-pushable/compare/v3.2.0...v3.2.1) (2023-07-04)


### Bug Fixes

* await returned promise in waitNext ([#59](https://github.com/alanshaw/it-pushable/issues/59)) ([d19e8ca](https://github.com/alanshaw/it-pushable/commit/d19e8caf02a98989bd4c42f60ccc8d616bdbe12b))

## [3.2.0](https://github.com/alanshaw/it-pushable/compare/v3.1.4...v3.2.0) (2023-07-03)


### Features

* add onEmpty function that resolves when the queue is empty ([#58](https://github.com/alanshaw/it-pushable/issues/58)) ([2bed38a](https://github.com/alanshaw/it-pushable/commit/2bed38ad1e477efc30c1a800c4de51813c8319d0))

## [3.1.4](https://github.com/alanshaw/it-pushable/compare/v3.1.3...v3.1.4) (2023-06-30)


### Dependencies

* **dev:** bump aegir from 37.12.1 to 39.0.13 ([#52](https://github.com/alanshaw/it-pushable/issues/52)) ([9399579](https://github.com/alanshaw/it-pushable/commit/939957932ed19a6558748e06ce50822b8d062b9b))

## [3.1.3](https://github.com/alanshaw/it-pushable/compare/v3.1.2...v3.1.3) (2023-04-18)


### Bug Fixes

* extend async generator ([#56](https://github.com/alanshaw/it-pushable/issues/56)) ([3cf8f5c](https://github.com/alanshaw/it-pushable/commit/3cf8f5c5dd6d7aac8e93f7bfe003b99caf9267ca))

## [3.1.2](https://github.com/alanshaw/it-pushable/compare/v3.1.1...v3.1.2) (2022-12-16)


### Documentation

* publish api docs ([#39](https://github.com/alanshaw/it-pushable/issues/39)) ([95adf08](https://github.com/alanshaw/it-pushable/commit/95adf08e789b8ca4617163026e091835aba39706))

## [3.1.1](https://github.com/alanshaw/it-pushable/compare/v3.1.0...v3.1.1) (2022-12-16)


### Trivial Changes

* **deps-dev:** bump it-all from 1.0.6 to 2.0.0 ([#34](https://github.com/alanshaw/it-pushable/issues/34)) ([1b50e5e](https://github.com/alanshaw/it-pushable/commit/1b50e5ed9211711530103db140cd401ecb2339fd))

## [3.1.0](https://github.com/alanshaw/it-pushable/compare/v3.0.0...v3.1.0) (2022-08-02)


### Features

* support Uint8ArrayLists in the same way as Uint8Arrays ([#30](https://github.com/alanshaw/it-pushable/issues/30)) ([7bae368](https://github.com/alanshaw/it-pushable/commit/7bae3688b1363954539b56de79c3a81ff53df59f))

## [3.0.0](https://github.com/alanshaw/it-pushable/compare/v2.0.2...v3.0.0) (2022-06-10)


### ⚠ BREAKING CHANGES

* `Uint8Array`s are expected by default, pass `objectMode: true` to push any other data types. If using TypeScript, use generics to define the data type.

### Features

* add readableLength property ([#27](https://github.com/alanshaw/it-pushable/issues/27)) ([f45aee3](https://github.com/alanshaw/it-pushable/commit/f45aee36e72e754b8a27dda48d3051c470aaa8e5))


### Trivial Changes

* fix flaky test ([#28](https://github.com/alanshaw/it-pushable/issues/28)) ([8c3dcc1](https://github.com/alanshaw/it-pushable/commit/8c3dcc1f8e64f2877317a835bb03545f7fa2dd53))

### [2.0.2](https://github.com/alanshaw/it-pushable/compare/v2.0.1...v2.0.2) (2022-06-08)


### Bug Fixes

* update aegir ([#24](https://github.com/alanshaw/it-pushable/issues/24)) ([9c1a478](https://github.com/alanshaw/it-pushable/commit/9c1a4783a536d90bcede5d29cc2d66d2a0d5321a))


### Trivial Changes

* update publish command ([#25](https://github.com/alanshaw/it-pushable/issues/25)) ([b82173f](https://github.com/alanshaw/it-pushable/commit/b82173f3c7d07581e1ece3ed3026ccbbf6c57056))

### [2.0.1](https://github.com/alanshaw/it-pushable/compare/v2.0.0...v2.0.1) (2022-01-13)


### Trivial Changes

* fix readme example ([#13](https://github.com/alanshaw/it-pushable/issues/13)) ([d4d7282](https://github.com/alanshaw/it-pushable/commit/d4d728275ba97977fd2004be749a57bbb74aebca))

## [2.0.0](https://github.com/alanshaw/it-pushable/compare/v1.4.2...v2.0.0) (2022-01-13)


### ⚠ BREAKING CHANGES

* switch to named exports, ESM only

### Features

* convert to typescript ([#12](https://github.com/alanshaw/it-pushable/issues/12)) ([49e0805](https://github.com/alanshaw/it-pushable/commit/49e080564a410a5f3475dfaa389ad8f0f1d8582c))


test.js
const test = require('ava')
const pipe = require('it-pipe')
const pushable = require('.')

const collect = async source => {
  const input = []
  for await (const value of source) input.push(value)
  return input
}

test('should push input slowly', async t => {
  const source = pushable()
  const input = [1, 2, 3]
  for (let i = 0; i < input.length; i++) {
    setTimeout(() => source.push(input[i]), i * 10)
  }
  setTimeout(() => source.end(), input.length * 10)
  const output = await pipe(source, collect)
  t.deepEqual(output, input)
})

test('should buffer input', async t => {
  const source = pushable()
  const input = [1, 2, 3]
  input.forEach(v => source.push(v))
  setTimeout(() => source.end())
  const output = await pipe(source, collect)
  t.deepEqual(output, input)
})

test('should buffer some inputs', async t => {
  const source = pushable()
  const input = [1, [2.1, 2.2, 2.3], 3, 4, 5, [6.1, 6.2, 6.3, 6.4], 7]
  for (let i = 0; i < input.length; i++) {
    setTimeout(() => {
      if (Array.isArray(input[i])) {
        input[i].forEach(v => source.push(v))
      } else {
        source.push(input[i])
      }
    }, i * 10)
  }
  setTimeout(() => source.end(), input.length * 10)
  const output = await pipe(source, collect)
  t.deepEqual(output, [].concat.apply([], input))
})

test('should allow end before start', async t => {
  const source = pushable()
  const input = [1, 2, 3]
  input.forEach(v => source.push(v))
  source.end()
  const output = await pipe(source, collect)
  t.deepEqual(output, input)
})

test('should end with error immediately', async t => {
  const source = pushable()
  const input = [1, 2, 3]
  input.forEach(v => source.push(v))
  source.end(new Error('boom'))
  const err = await t.throwsAsync(pipe(source, collect))
  t.deepEqual(err.message, 'boom')
})

test('should end with error in the middle', async t => {
  const source = pushable()
  const input = [1, new Error('boom'), 3]
  for (let i = 0; i < input.length; i++) {
    setTimeout(() => {
      if (input[i] instanceof Error) {
        source.end(input[i])
      } else {
        source.push(input[i])
      }
    }, i * 10)
  }
  setTimeout(() => source.end(), input.length * 10)
  const err = await t.throwsAsync(pipe(source, collect))
  t.deepEqual(err.message, 'boom')
})

test('should allow end without push', async t => {
  const source = pushable()
  const input = []
  source.end()
  const output = await pipe(source, collect)
  t.deepEqual(output, input)
})

test('should allow next after end', async t => {
  const source = pushable()
  const input = [1]
  source.push(input[0])
  let next = await source.next()
  t.falsy(next.done)
  t.is(next.value, input[0])
  source.end()
  next = await source.next()
  t.true(next.done)
  next = await source.next()
  t.true(next.done)
})

test.cb('should call onEnd', t => {
  const source = pushable(() => t.end())
  const input = [1, 2, 3]
  for (let i = 0; i < input.length; i++) {
    setTimeout(() => source.push(input[i]), i * 10)
  }
  setTimeout(() => source.end(), input.length * 10)
  pipe(source, collect)
})

test.cb('should call onEnd if passed in options object', t => {
  const source = pushable({ onEnd: () => t.end() })
  const input = [1, 2, 3]
  for (let i = 0; i < input.length; i++) {
    setTimeout(() => source.push(input[i]), i * 10)
  }
  setTimeout(() => source.end(), input.length * 10)
  pipe(source, collect)
})

test.cb('should call onEnd even if not piped', t => {
  const source = pushable(() => t.end())
  source.end()
})

test.cb('should call onEnd with error', t => {
  const source = pushable(err => {
    t.is(err.message, 'boom')
    t.end()
  })
  setTimeout(() => source.end(new Error('boom')), 10)
  pipe(source, collect).catch(() => {})
})

test.cb('should call onEnd on return before end', t => {
  const input = [1, 2, 3, 4, 5]
  const max = 2
  const output = []

  const source = pushable(() => {
    t.deepEqual(output, input.slice(0, max))
    t.end()
  })

  input.forEach((v, i) => setTimeout(() => source.push(v), i * 10))
  setTimeout(() => source.end(), input.length * 10)

  ;(async () => {
    let i = 0
    for await (const value of source) {
      output.push(value)
      i++
      if (i === max) break
    }
  })()
})

test.cb('should call onEnd by calling return', t => {
  const input = [1, 2, 3, 4, 5]
  const max = 2
  const output = []

  const source = pushable(() => {
    t.deepEqual(output, input.slice(0, max))
    t.end()
  })

  input.forEach((v, i) => setTimeout(() => source.push(v), i * 10))
  setTimeout(() => source.end(), input.length * 10)

  ;(async () => {
    let i = 0
    while (i !== max) {
      i++
      const { value } = await source.next()
      output.push(value)
    }
    source.return()
  })()
})

test.cb('should call onEnd once', t => {
  const input = [1, 2, 3, 4, 5]

  let count = 0
  const source = pushable(() => {
    count++
    t.is(count, 1)
    setTimeout(() => t.end(), 50)
  })

  input.forEach((v, i) => setTimeout(() => source.push(v), i * 10))

  ;(async () => {
    await source.next()
    source.return()
    source.next()
  })()
})

test.cb('should call onEnd by calling throw', t => {
  const input = [1, 2, 3, 4, 5]
  const max = 2
  const output = []

  const source = pushable(err => {
    t.is(err.message, 'boom')
    t.deepEqual(output, input.slice(0, max))
    t.end()
  })

  input.forEach((v, i) => setTimeout(() => source.push(v), i * 10))
  setTimeout(() => source.end(), input.length * 10)

  ;(async () => {
    let i = 0
    while (i !== max) {
      i++
      const { value } = await source.next()
      output.push(value)
    }
    source.throw(new Error('boom'))
  })()
})


test/fifo.spec.ts
import { expect } from 'aegir/chai'
import { FIFO } from '../src/fifo.js'

describe('fifo', () => {
  it('basic', function () {
    const q = new FIFO()
    const values: any[] = [
      1,
      4,
      4,
      0,
      null,
      {},
      Math.random(),
      '',
      'hello',
      9,
      1,
      4,
      5,
      6,
      7,
      null,
      null,
      0,
      0,
      15,
      52.2,
      null
    ]

    expect(q.shift()).to.be.undefined()
    expect(q.isEmpty()).to.be.true()
    expect(q.size).to.equal(0)

    for (const value of values) {
      q.push(value)
    }

    expect(q.size).to.equal(values.length)

    while (!q.isEmpty()) {
      expect(q.shift()).to.deep.equal(values.shift())
      expect(q.size).to.equal(values.length)
    }

    expect(q.shift()).to.be.undefined()
    expect(q.isEmpty()).to.be.true()
  })

  it('long length', function () {
    const q = new FIFO<number>()

    const len = 0x8f7
    for (let i = 0; i < len; i++) {
      q.push(i)
    }

    expect(q.size).to.equal(len)

    let shifts = 0
    while (!q.isEmpty()) {
      q.shift()
      shifts++
    }

    expect(shifts).to.equal(len)
    expect(q.size).to.equal(0)
  })

  it('clear', function () {
    const q = new FIFO<string>()

    q.push('a')
    q.push('a')
    q.clear()

    expect(q.shift()).to.be.undefined()
    expect(q.size).to.equal(0)

    for (let i = 0; i < 50; i++) {
      q.push('a')
    }

    q.clear()

    expect(q.shift()).to.be.undefined()
    expect(q.size).to.equal(0)
  })

  it('basic length', function () {
    const q = new FIFO<string>()

    q.push('a')
    expect(q.size).to.equal(1)

    q.push('a')
    expect(q.size).to.equal(2)

    q.shift()
    expect(q.size).to.equal(1)

    q.shift()
    expect(q.size).to.equal(0)

    q.shift()
    expect(q.size).to.equal(0)
  })

  it('peek', function () {
    const q = new FIFO<string>()

    q.push('a')
    expect(q.size).to.equal(1)
    expect(q.peek()).to.deep.equal('a')
    expect(q.peek()).to.deep.equal('a')

    q.push('b')
    expect(q.size).to.equal(2)
    expect(q.peek()).to.deep.equal('a')
    expect(q.peek()).to.deep.equal('a')

    expect(q.shift()).to.deep.equal('a')
    expect(q.peek()).to.deep.equal('b')
    expect(q.peek()).to.deep.equal('b')

    expect(q.shift()).to.deep.equal('b')
    expect(q.peek()).to.be.undefined()
    expect(q.peek()).to.be.undefined()
  })

  it('peek edgecase', function () {
    const q = new FIFO<string>({ splitLimit: 4 })

    q.push('a')
    q.push('b')
    q.push('c')
    q.push('d')
    q.push('e')

    expect(q.peek()).to.equal(q.shift())
    expect(q.peek()).to.equal(q.shift())
    expect(q.peek()).to.equal(q.shift())
    expect(q.peek()).to.equal(q.shift())
    expect(q.peek()).to.equal(q.shift())
    expect(q.peek()).to.equal(q.shift())
  })

  it('invalid hwm', function () {
    expect(() => new FIFO({ splitLimit: 3 })).to.throw()
  })
})


index.js
const FIFO = require('fast-fifo')

module.exports = (options) => {
  options = options || {}
  let onEnd

  if (typeof options === 'function') {
    onEnd = options
    options = {}
  } else {
    onEnd = options.onEnd
  }

  let buffer = new FIFO()
  let pushable, onNext, ended

  const waitNext = () => {
    if (!buffer.isEmpty()) {
      const next = buffer.shift()
      if (next.error) throw next.error
      return next
    }

    if (ended) return { done: true }

    return new Promise((resolve, reject) => {
      onNext = next => {
        onNext = null
        if (next.error) {
          reject(next.error)
        } else {
          resolve(next)
        }
        return pushable
      }
    })
  }

  const bufferNext = next => {
    if (onNext) return onNext(next)
    buffer.push(next)
    return pushable
  }

  const bufferError = err => {
    buffer = new FIFO()
    if (onNext) return onNext({ error: err })
    buffer.push({ error: err })
    return pushable
  }

  const push = value => {
    if (ended) return pushable
    return bufferNext({ done: false, value })
  }
  const end = err => {
    if (ended) return pushable
    ended = true
    return err ? bufferError(err) : bufferNext({ done: true })
  }
  const _return = () => {
    buffer = new FIFO()
    end()
    return { done: true }
  }
  const _throw = err => {
    end(err)
    return { done: true }
  }

  pushable = {
    [Symbol.asyncIterator] () { return this },
    next: waitNext,
    return: _return,
    throw: _throw,
    push,
    end
  }

  if (!onEnd) return pushable

  const _pushable = pushable

  pushable = {
    [Symbol.asyncIterator] () { return this },
    next () {
      return _pushable.next()
    },
    throw (err) {
      _pushable.throw(err)
      if (onEnd) {
        onEnd(err)
        onEnd = null
      }
      return { done: true }
    },
    return () {
      _pushable.return()
      if (onEnd) {
        onEnd()
        onEnd = null
      }
      return { done: true }
    },
    push,
    end (err) {
      _pushable.end(err)
      if (onEnd) {
        onEnd(err)
        onEnd = null
      }
      return pushable
    }
  }

  return pushable
}


src/fifo.ts
// ported from https://www.npmjs.com/package/fast-fifo

class FixedFIFO<T> {
  public buffer: Array<T | undefined>
  private readonly mask: number
  private top: number
  private btm: number
  public next: FixedFIFO<T> | null

  constructor (hwm: number) {
    if (!(hwm > 0) || ((hwm - 1) & hwm) !== 0) {
      throw new Error('Max size for a FixedFIFO should be a power of two')
    }

    this.buffer = new Array(hwm)
    this.mask = hwm - 1
    this.top = 0
    this.btm = 0
    this.next = null
  }

  clear (): void {
    this.top = this.btm = 0
    this.next = null
    this.buffer.fill(undefined)
  }

  push (data: T): boolean {
    if (this.buffer[this.top] !== undefined) {
      return false
    }

    this.buffer[this.top] = data
    this.top = (this.top + 1) & this.mask

    return true
  }

  shift (): T | undefined {
    const last = this.buffer[this.btm]

    if (last === undefined) {
      return undefined
    }

    this.buffer[this.btm] = undefined
    this.btm = (this.btm + 1) & this.mask

    return last
  }

  peek (): T | undefined {
    return this.buffer[this.btm]
  }

  isEmpty (): boolean {
    return this.buffer[this.btm] === undefined
  }
}

export interface FIFOOptions<T> {
  /**
   * When the queue reaches this size, it will be split into head/tail parts
   */
  splitLimit?: number

  calculateSize?(value?: T): number
}

function defaultCalculateSize (obj: any): number {
  if (obj?.byteLength != null) {
    return obj.byteLength
  }

  return 1
}

export class FIFO<T> {
  public size: number
  private readonly hwm: number
  private head: FixedFIFO<T>
  private tail: FixedFIFO<T>
  private readonly calculateSize: (value?: T) => number

  constructor (options: FIFOOptions<T> = {}) {
    this.hwm = options.splitLimit ?? 16
    this.head = new FixedFIFO<T>(this.hwm)
    this.tail = this.head
    this.size = 0
    this.calculateSize = options.calculateSize ?? defaultCalculateSize
  }

  clear (): void {
    this.head = this.tail
    this.head.clear()
    this.size = 0
  }

  push (val: T): void {
    this.size += this.calculateSize(val)

    if (!this.head.push(val)) {
      const prev = this.head
      this.head = prev.next = new FixedFIFO<T>(2 * this.head.buffer.length)
      this.head.push(val)
    }
  }

  shift (): T | undefined {
    let val = this.tail.shift()

    if (val === undefined && (this.tail.next != null)) {
      const next = this.tail.next
      this.tail.next = null
      this.tail = next
      val = this.tail.shift()
    }

    if (this.size > 0) {
      this.size -= this.calculateSize(val)
    }

    return val
  }

  peek (): T | undefined {
    const val = this.tail.peek()

    if (val === undefined && this.tail.next != null) {
      return this.tail.next.peek()
    }

    return val
  }

  isEmpty (): boolean {
    return this.head.isEmpty()
  }
}


src/index.ts
/**
 * @packageDocumentation
 *
 * An iterable that you can push values into.
 *
 * @example
 *
 * ```js
 * import { pushable } from 'it-pushable'
 *
 * const source = pushable()
 *
 * setTimeout(() => source.push('hello'), 100)
 * setTimeout(() => source.push('world'), 200)
 * setTimeout(() => source.end(), 300)
 *
 * const start = Date.now()
 *
 * for await (const value of source) {
 *   console.log(`got "${value}" after ${Date.now() - start}ms`)
 * }
 * console.log(`done after ${Date.now() - start}ms`)
 *
 * // Output:
 * // got "hello" after 105ms
 * // got "world" after 207ms
 * // done after 309ms
 * ```
 *
 * @example
 *
 * ```js
 * import { pushableV } from 'it-pushable'
 * import all from 'it-all'
 *
 * const source = pushableV()
 *
 * source.push(1)
 * source.push(2)
 * source.push(3)
 * source.end()
 *
 * console.info(await all(source))
 *
 * // Output:
 * // [ [1, 2, 3] ]
 * ```
 *
 * ### Backpressure
 * This module supports backpressure by returning a promise to any `push` or `end` call.
 * These promises will resolve when the `push`ed value has been consumed (in the case of `.push`) or when the consumer has iterated over all values in the pushable (in the case of `.end`).
 *
 * If you do not wish to wait and instead buffer as much data as possible, do not await the result of `.push`
 *
 * @example
 *
 * ```js
 * import { pushable } from 'it-pushable'
 *
 * const source = pushable()
 *
 * // wait for the value to be consumed
 * await source.push(5)
 *
 * // do not wait for the value to be consumed
 * void source.push(5)
 *
 * // only wait 10ms for the value to be consumed
 * await source.push(5, {
 *   signal: AbortSignal.timeout(10)
 * })
```
 */

import defer, { type DeferredPromise } from 'p-defer'
import { raceSignal, type RaceSignalOptions } from 'race-signal'
import { PFIFO } from './p-fifo.js'

export interface AbortOptions {
  signal?: AbortSignal
}

export interface PushableOptions {
  /**
   * A function called after *all* values have been yielded from the iterator
   * (including buffered values). In the case when the iterator is ended with an
   * error it will be passed the error as a parameter.
   */
  onEnd?(err?: Error): void
}

export interface Pushable<T> extends AsyncGenerator<T, void, unknown> {
  /**
   * End the iterable after all values in the buffer (if any) have been yielded.
   * If an error is passed the buffer is cleared immediately and the next
   * iteration will throw the passed error
   *
   * If `.next` has not been called on this iterable, `.end` will return
   * immediately. Otherwise it will wait for the final value to be consumed
   * before resolving, unless the passed signal emits an `abort` event before
   * then.
   */
  end(err?: Error, options?: AbortOptions & RaceSignalOptions): Promise<void>

  /**
   * Push a value into the iterable. Values are yielded from the iterable in the
   * order they are pushed. Values not yet consumed from the iterable are
   * buffered.
   *
   * When the returned promise resolves, the pushed value will have been
   * consumed.
   */
  push(value: T, options?: AbortOptions & RaceSignalOptions): Promise<void>

  /**
   * This property contains the number of bytes (or objects) in the queue ready
   * to be read.
   *
   * If objects pushed into this queue have a `.byteLength` property, this is
   * ths total number of bytes in the queue, otherwise it is the number of
   * objects in the queue
   */
  readableLength: number
}

export interface PushableV<T> extends AsyncGenerator<T[], void, unknown> {
  /**
   * End the iterable after all values in the buffer (if any) have been yielded.
   * If an error is passed the buffer is cleared immediately and the next
   * iteration will throw the passed error
   *
   * If `.next` has not been called on this iterable, `.end` will return
   * immediately. Otherwise it will wait for the final value to be consumed
   * before resolving, unless the passed signal emits an `abort` event before
   * then.
   */
  end(err?: Error, options?: AbortOptions & RaceSignalOptions): Promise<void>

  /**
   * Push a value into the iterable. Values are yielded from the iterable in the
   * order they are pushed. Values not yet consumed from the iterable are
   * buffered.
   *
   * When the returned promise resolves, the pushed value will have been
   * consumed.
   */
  push(value: T, options?: AbortOptions & RaceSignalOptions): Promise<void>

  /**
   * Push a list of values into the iterable in-order. Values are yielded from
   * the iterable in the order they are pushed. Values not yet consumed from the
   * iterable are buffered.
   *
   * When the returned promise resolves, all pushed values will have been
   * consumed.
   */
  pushV (values: T[], options?: AbortOptions & RaceSignalOptions): Promise<void>

  /**
   * This property contains the number of bytes (or objects) in the queue ready
   * to be read.
   *
   * If objects pushed into this queue have a `.byteLength` property, this is
   * ths total number of bytes in the queue, otherwise it is the number of
   * objects in the queue
   */
  readableLength: number
}

abstract class AbstractPushable<T> {
  protected readonly errorPromise: DeferredPromise<IteratorReturnResult<void>>
  protected readonly returnPromise: DeferredPromise<IteratorReturnResult<undefined>>
  protected fifo: PFIFO<IteratorResult<T, void>>
  protected onEnd?: (err?: Error) => void
  protected piped: boolean
  protected ended: boolean

  constructor (options?: PushableOptions) {
    this.errorPromise = defer()
    this.returnPromise = defer()
    this.fifo = new PFIFO()
    this.onEnd = options?.onEnd
    this.piped = false
    this.ended = false
  }

  [Symbol.asyncIterator] (): this {
    return this
  }

  async next (): Promise<IteratorResult<T, void>> {
    this.piped = true

    try {
      const result = await Promise.race([
        this.errorPromise.promise,
        this.returnPromise.promise,
        (async () => {
          if (this.ended && this.fifo.isEmpty()) {
            const result: IteratorReturnResult<void> = {
              done: true,
              value: undefined
            }

            return result
          }

          // if this.fifo is empty, this will wait for either a value to be
          // pushed, or for .throw, .return or .end to be called
          return this.getValues()
        })()
      ])

      if (result.done === true) {
        this.onEnd?.()
        this.onEnd = undefined
      }

      return result
    } catch (err: any) {
      this.onEnd?.(err)
      this.onEnd = undefined

      throw err
    }
  }

  async return (): Promise<IteratorReturnResult<void>> {
    this.ended = true
    this.returnPromise.resolve({ done: true, value: undefined })
    this.fifo = new PFIFO()

    this.onEnd?.()
    this.onEnd = undefined

    return {
      done: true,
      value: undefined
    }
  }

  async throw (err: Error): Promise<IteratorReturnResult<undefined>> {
    this.ended = true
    this.errorPromise.reject(err)
    this.fifo = new PFIFO()

    this.onEnd?.(err)
    this.onEnd = undefined

    return {
      done: true,
      value: undefined
    }
  }

  async end (err?: Error, options?: AbortOptions & RaceSignalOptions): Promise<void> {
    this.ended = true

    if (err != null) {
      await this.throw(err)
      return
    }

    await new Promise<void>((resolve, reject) => {
      // settle promise in the microtask queue to give consumers a chance to
      // await after calling .push
      queueMicrotask(() => {
        if (!this.piped && this.fifo.isEmpty()) {
          // never pushed or piped
          this.onEnd?.(err)
          this.onEnd = undefined

          resolve()
          return
        }

        raceSignal(
          this.fifo.push({ done: true, value: undefined }),
          options?.signal,
          options
        )
          .then(() => { resolve() }, (err) => { reject(err) })
      })
    })
  }

  get readableLength (): number {
    return this.fifo.size
  }

  protected abstract getValues (): Promise<IteratorResult<T, void>>
}

class ItPushable<T> extends AbstractPushable<T> implements Pushable<T> {
  protected async getValues (): Promise<IteratorResult<T, void>> {
    return this.fifo.shift()
  }

  async push (value: T, options?: AbortOptions & RaceSignalOptions): Promise<void> {
    if (this.ended) {
      throw new Error('pushable has already ended')
    }

    await raceSignal(
      this.fifo.push({ done: false, value }),
      options?.signal,
      options
    )
  }
}

class ItPushableV<T> extends AbstractPushable<T[]> implements PushableV<T> {
  protected async getValues (): Promise<IteratorResult<T[], void>> {
    const promises: Array<Promise<IteratorResult<T[]>>> = []

    while (!this.fifo.isEmpty()) {
      promises.push(this.fifo.shift())
    }

    const output: IteratorResult<T[]> = {
      done: false,
      value: []
    }

    for (const result of await Promise.all(promises)) {
      if (result.done === true) {
        if (output.value.length > 0) {
          // we have some values, return them but empty the iterator
          // so we return { done: true } on the subsequent calls to
          // .next()
          this.fifo = new PFIFO()
          this.ended = true
          break
        }

        // did not have any values, just return done result
        return {
          done: true,
          value: undefined
        }
      }

      if (result.value != null) {
        output.value.push(...result.value)
      }
    }

    if (output.value.length === 0) {
      return {
        done: true,
        value: undefined
      }
    }

    return output
  }

  async push (value: T, options?: AbortOptions & RaceSignalOptions): Promise<void> {
    await this.pushV([value], options)
  }

  async pushV (values: T[], options?: AbortOptions & RaceSignalOptions): Promise<void> {
    if (this.ended) {
      throw new Error('pushable has already ended')
    }

    await raceSignal(
      this.fifo.push({ done: false, value: values }),
      options?.signal,
      options
    )
  }
}

export function pushable<T> (options?: PushableOptions): Pushable<T>
export function pushable<T> (options?: PushableOptions): Pushable<T> {
  return new ItPushable<T>(options)
}

export function pushableV<T> (options?: PushableOptions): PushableV<T>
export function pushableV<T> (options?: PushableOptions): PushableV<T> {
  return new ItPushableV<T>(options)
}
